clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -116,9 +116,9 @@
     public void seekToEnd(Collection<TopicPartition> partitions);
 
     /**
-     * @see KafkaConsumer#position(TopicPartition)
+     * @see KafkaConsumer#position(TopicPartition, long)
      */
-    public long position(TopicPartition partition);
+    public long position(TopicPartition partition, long timeout);
```

Comment:
OK. I think I understand the problem a little better after reading @guozhangwang's comment on the JIRA. I'm thinking it might be sufficient to handle this case specially. When automatic assignment is used, there really should be no reason we block on partitions that we know don't exist. From a high level, I'm wondering if we can consult the local metadata cache and if we see that the partition number is greater than the number of partitions, then 1) if we're inside `position()`, we raise an exception, and 2) if we're inside `poll()`, we effectively pause that partition and request another rebalance.